thetis/callback.py

```diff
@@ -481,3 +481,49 @@ def __call__(self):
             field_vals.append(np.reshape(self._evaluate_field(field_name), (ndetectors, -1)))
 
         return np.hstack(field_vals)
+
+
+class FunctionalCallback(DiagnosticCallback):
+    """Base class for callbacks that evaluate a functional involving integrals in both time and space. This callback
+    can also be used to assemble time dependent objective functionals for adjoint simulations. Time integration is
+    achieved using the trapezium rule."""
+    variable_names = ['current integral', 'objective value']
+
+    def __init__(self, scalar_callback, solver_obj, **kwargs):
+        """
+        Creates error comparison check callback object
+
+        :arg scalar_callback: Python function that takes the solver object as an argument and
+            returns a list of values of an objective functional.
+        :arg solver_obj: Thetis solver object
+        :arg **kwargs: any additional keyword arguments, see DiagnosticCallback
+        """
+        kwargs.setdefault('export_to_hdf5', False)
+        kwargs.setdefault('append_to_log', False)
+        super(FunctionalCallback, self).__init__(solver_obj, **kwargs)
+        self.scalar_callback = scalar_callback      # Evaluate functional
+        self.objective_value = [scalar_callback()]  # Store functional value
+        self.dt = solver_obj.options.timestep
+        self.t = solver_obj.simulation_time
+
+    def __call__(self):
+        value = self.scalar_callback()
+        if self.t > 0.5 * self.dt:              # Don't store the first value twice
+            self.objective_value.append(value)
+        self.t += self.dt
+
+        return value, self.objective_value
```

Comment:
Do you have to return objective_value. I think it breaks using hdf5 output for this DiagnosticCallback, as it will be of a different length after each call.